etl/steps/data/meadow/climate/2024-11-19/total_precipitation.py

```diff
@@ -0,0 +1,167 @@
+    temp_country["World"] = global_mean
+
+    # Initialize a list to keep track of small countries where precipitation data extraction fails.
+    small_countries = []
+
+    # Iterate over each row in the shapefile data.
+    for i in tqdm(range(shapefile.shape[0])):
+        # Extract the data for the current row.
+        geometry = shapefile.iloc[i]["geometry"]
+        country_name = shapefile.iloc[i]["WB_NAME"]
+
+        try:
+            # Clip to the bounding box for the country's shape to significantly improve performance.
+            xmin, ymin, xmax, ymax = geometry.bounds
+            clip = da.rio.clip_box(minx=xmin, miny=ymin, maxx=xmax, maxy=ymax)
+
+            # Clip data to the country's shape.
+            # NOTE: if memory is an issue, we could use `from_disk=True` arg
+            clip = clip.rio.clip([mapping(geometry)], shapefile.crs)
+
+            # Calculate weights based on latitude to account for area distortion in latitude-longitude grids.
+            weights = np.cos(np.deg2rad(clip.latitude))
+            weights.name = "weights"
+
+            # Apply the weights to the clipped precipitation data.
+            clim_month_weighted = clip.weighted(weights)
+
+            # Calculate the weighted mean precipitation for the country.
+            # country_weighted_mean = clim_month_weighted.mean(dim=["longitude", "latitude"]).values
+            country_sum = clim_month_weighted.mean(dim=["longitude", "latitude"]).values
+
+            # Store the calculated mean precipitation in the dictionary with the country's name as the key.
+            temp_country[country_name] = country_sum
+
+            # Clean up the memory
+            del clip
+            del weights
+            del clim_month_weighted
+
+        except (NoDataInBounds, OneDimensionalRaster):
+            log.info(
+                f"No data was found in the specified bounds for {country_name}."
+            )  # If an error occurs (usually due to small size of the country), add the country's name to the small_countries list.  # If an error occurs (usually due to small size of the country), add the country's name to the small_countries list.
+            small_countries.append(shapefile.iloc[i]["WB_NAME"])
+
+    # Log information about countries for which precipitation data could not be extracted.
+    log.info(
+        f"It wasn't possible to extract precipitation data for {len(small_countries)} small countries as they are too small for the resolution of the Copernicus data."
+    )
+    # Define the start and end dates
+    da["date"] = pd.to_datetime(da["date"].astype(str), format="%Y%m%d")
+
+    # Now you can access the 'dt' accessor
+    start_time = da["date"].min().dt.date.astype(str).item()
+    end_time = da["date"].max().dt.date.astype(str).item()
+
+    # Generate a date range from start_time to end_time with monthly frequency
+    month_middles = pd.date_range(start=start_time, end=end_time, freq="MS") + pd.offsets.Day(14)
```

Comment:
I wanted the monthly data to be in the middle of the month on grapher. Though I am not actually plotting monthly data atm I think I'll keep it in case I do make those charts in the future!